DESIGN.md

```diff
@@ -0,0 +1,773 @@
+socket, to the kernel using the submission queue. The kernel then performs the
+operation. On completion, the kernel returns the operation results via the
+completion queue and notifies the process. The `io_uring_enter` syscall flushes
+the submission queue and acquires any pending completion events. Upon request,
+this syscall may block the thread waiting for a minimum number of completion
+events. Both queues use memory shared between the process and the kernel and
+synchronize with atomic operations.
+
+Operation futures provide an asynchronous cancel function, enabling the caller
+to await on a clean cancellation.
+
+```rust
+let accept = tcp_listener.accept();
+
+tokio::select! {
+    (tcp_stream, addr) = &mut accept => { ... }
+    _ = tokio::time::sleep(Duration::from_millis(100)) => {
+        // Accept timed out, cancel the in-flight accept.
+        match accept.cancel().await {
+            Ok(_) => { ... } // operation canceled gracefully
+            Err(Completed(tcp_stream, addr)) => {
+                // The operation completed between the timeout
+                // and cancellation request.
+            }
+            Err(_) => { ... }
+        }
+    }
+}
+```
+
+In practice, operation timeouts will deserve a dedicated API to handle the
+boilerplate.
+
+```rust
+let (tcp_stream, addr) = tcp_listener
+    .accept()
+    .timeout(Duration::from_millis(100))
+    .await?
+```
+
+The cancel and timeout function are inherent methods on operation future types.
+
+If the operation's future drops before the operation completes, the runtime will
+submit a cancellation request for the operation. However, cancellation is
+asynchronous and best-effort, the operation may still complete. In that case,
+the runtime discards the operation result.
+
+The queue's single-producer characteristic optimizes for a single thread to own
+a given submission queue. Supporting a multi-threaded runtime requires either
+synchronizing pushing onto the submission queue or creating a queue pair per
+thread. The tokio-uring runtime uses an isolated thread-per-core model,
+differing from Tokio's other runtimes. Unlike Tokio's primary multi-threaded
+runtime, there is no work-stealing. Tasks remain on the thread that spawned them
+for the duration of their lifecycle. Each runtime thread will own a dedicated
+submission and completion queue pair, and operations are submitted using the
+submission queue associated with the current thread. Operation completion
+futures will not implement `Send`, guaranteeing that they remain on the thread
+to receive the operation result. Interestingly, the resources, e.g.,
+`TcpListener`, can be `Send` as long as they do not hold operation futures
+internally. It is possible to have two operations in-flight from a single
+resource associated with different queues and threads.
+
+Because operations are not `Send`, tasks awaiting these operations are also not
```

Comment:
Nit: I "not send" should have a consistent symbol across the doc, and I personally think using `!Send` makes it easier to read.